File: HoTT-Model/src/main/java/gde/model/voice/VoiceData.java
package gde.model.voice;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.Serializable;
import java.nio.file.Files;
import java.util.Arrays;

import javax.sound.sampled.AudioFileFormat;
import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioFormat.Encoding;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;

import org.apache.commons.io.IOUtils;

/**
 * Representation of a single announcement in a voice data file (.vdf).
 *
 * @author oliver.treichel@gmx.de
 */
public class VoiceData implements Serializable {
    private static final long serialVersionUID = 1L;

    /** Default audio format 11 kHz 16-bit signed PCM mono */
    public static final AudioFormat AUDIO_FORMAT = new AudioFormat(Encoding.PCM_SIGNED, 11025, 16, 1, 2, 11025, false);

    public static VoiceData readSoundFile(final File soundFile) throws UnsupportedAudioFileException, IOException {
        // read from file
        final AudioInputStream sourceAudioStream = AudioSystem.getAudioInputStream(soundFile);
        final AudioFormat sourceFormat = sourceAudioStream.getFormat();
        final float rate = sourceFormat.getSampleRate();
        final int channels = sourceFormat.getChannels();

        // convert to from MP3 or OGG to PCM
        final AudioFormat pcmFormat = new AudioFormat(Encoding.PCM_SIGNED, rate, 16, channels, channels * 2, rate, false);
        final AudioInputStream pcmAudioStream = sourceFormat.getEncoding() == Encoding.PCM_SIGNED ? sourceAudioStream
                : AudioSystem.getAudioInputStream(pcmFormat, sourceAudioStream);

        // convert sample rate and channels
        final AudioInputStream targetAudioStream = pcmFormat.matches(AUDIO_FORMAT) ? pcmAudioStream
                : AudioSystem.getAudioInputStream(AUDIO_FORMAT, pcmAudioStream);

        // encode to ADPCM
        final InputStream encodedStream = ADPCMCodec.encode(targetAudioStream);

        final String fileName = soundFile.getName();
        final int dot = fileName.lastIndexOf(".");
        return new VoiceData(fileName.substring(0, dot), IOUtils.toByteArray(encodedStream));
    }

    private String name;
    private final byte[] data;

    public VoiceData(final String name, final byte[] data) {
        this.name = name;
        this.data = data;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        final VoiceData other = (VoiceData) obj;
        if (!Arrays.equals(data, other.data)) return false;
        if (name == null) {
            if (other.name != null) return false;
        } else if (!name.equals(other.name)) return false;
        return true;
    }

    public AudioInputStream getAudioInputStream() throws IOException {
        return new AudioInputStream(getPcmInputStream(), AUDIO_FORMAT, data.length * 2);
    }

    public String getName() {
        return name;
    }

    public byte[] getPcmData() throws IOException {
        return ADPCMCodec.decode(data);
    }

    public InputStream getPcmInputStream() throws IOException {
        return ADPCMCodec.decode(getRawInputStream());
    }

    public byte[] getRawData() {
        return data;
    }

    public InputStream getRawInputStream() {
        return new ByteArrayInputStream(getRawData());
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Arrays.hashCode(data);
        result = prime * result + (name == null ? 0 : name.hashCode());
        return result;
    }

    public void play() throws LineUnavailableException, InterruptedException, IOException {
        Player.play(AUDIO_FORMAT, getPcmInputStream());
    }

    public void setName(final String name) {
        this.name = name;
    }

    public void writeVox(final File voxFile) throws IOException {
        Files.write(voxFile.toPath(), getRawData());
    }

    public void writeWav(final File wavFile) throws IOException {
        AudioSystem.write(getAudioInputStream(), AudioFileFormat.Type.WAVE, wavFile);
    }
}

File: HoTT-Voice/src/main/kotlin/de/treichels/hott/voice/VoiceData.kt
package de.treichels.hott.voice

import org.apache.commons.io.IOUtils
import java.io.*
import java.nio.file.Files
import java.util.*
import javax.sound.sampled.*
import javax.sound.sampled.AudioFormat.Encoding

/**
 * Representation of a single announcement in a voice data file (.vdf).
 *
 * @author oliver.treichel@gmx.de
 */
class VoiceData(name: String, val rawData: ByteArray) : Serializable, ObservableBase() {
    companion object {
        /** Default audio format 11 kHz 16-bit signed PCM mono  */
        private val audioFormat = AudioFormat(Encoding.PCM_SIGNED, 11025f, 16, 1, 2, 11025f, false)

        fun forStream(sourceAudioStream: AudioInputStream, name: String, volume: Double = 1.0): VoiceData {
            // read from stream
            val sourceFormat = sourceAudioStream.format
            val rate = sourceFormat.sampleRate
            val channels = sourceFormat.channels

            // convert to from MP3 or OGG to PCM
            val pcmFormat = AudioFormat(Encoding.PCM_SIGNED, rate, 16, channels, channels * 2, rate, false)
            val pcmAudioStream = if (sourceFormat.encoding === Encoding.PCM_SIGNED)
                sourceAudioStream
            else
                AudioSystem.getAudioInputStream(pcmFormat, sourceAudioStream)

            // convert sample rate and channels
            val targetAudioStream = if (pcmFormat.matches(audioFormat))
                pcmAudioStream
            else
                AudioSystem.getAudioInputStream(audioFormat, pcmAudioStream)

            // encode to ADPCM
            val encodedStream = ADPCMCodec.encode(targetAudioStream, volume)
            return VoiceData(name, IOUtils.toByteArray(encodedStream))
        }

        fun forFile(soundFile: File, volume: Double = 1.0): VoiceData {
            // read from file
            val sourceAudioStream = AudioSystem.getAudioInputStream(soundFile)
            val fileName = soundFile.name
            val dot = fileName.lastIndexOf(".")
            val name = fileName.substring(0, dot)
            return forStream(sourceAudioStream, name, volume)
        }
    }
    
    var name = name
        set(name) {
            field = if (name.length > 17) name.substring(0, 18) else name
            invalidate()
        }

    val audioInputStream: AudioInputStream
        get() = AudioInputStream(pcmInputStream, audioFormat, 2L * rawData.size.toLong())

    val pcmData: ByteArray
        get() = ADPCMCodec.decode(rawData)

    val pcmInputStream: InputStream
        get() = ADPCMCodec.decode(rawInputStream)

    val rawInputStream: InputStream
        get() = ByteArrayInputStream(rawData)

    fun play() {
        try {
            Player.play(audioFormat, pcmInputStream)
        } catch (e: LineUnavailableException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun toString(): String {
        return String.format("VoiceData [name=%s]", this.name)
    }


    fun writeVox(voxFile: File) {
        Files.write(voxFile.toPath(), rawData)
    }


    fun writeWav(wavFile: File) {
        AudioSystem.write(audioInputStream, AudioFileFormat.Type.WAVE, wavFile)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as VoiceData

        if (!Arrays.equals(rawData, other.rawData)) return false
        if (name != other.name) return false

        return true
    }

    override fun hashCode(): Int {
        var result = Arrays.hashCode(rawData)
        result = 31 * result + (name.hashCode())
        return result
    }

    fun clone(): VoiceData = VoiceData(this.name, this.rawData)
}
